chore(chart): helm chart for dynamic-nfs-provisioner

rahulkrishnanfs opened the pull request:
Signed-off-by: Rahul krishnan R A <rahulkrishnanfs@gmail.com>

## Pull Request template

**Why is this PR required? What issue does it fix?**:
Chart for the dynamic-nfs-provisioner to make the deployment more generic
**What this PR does?**:
 Helm chart for dynamic-nfs-provisioner
**Does this PR require any upgrade changes?**:
No
**If the changes in this PR are manually verified, list down the scenarios covered:**:
Install the chart to the desired namespace using helm 

**Checklist:**
- [x] Fixes #15 
- [x] PR Title follows the convention of  `<type>(<scope>): <subject>`
- [ ] Has the change log section been updated? 
- [ ] Commit has unit tests
- [ ] Commit has integration tests
- [ ] (Optional) Are upgrade changes included in this PR? If not, mention the issue/PR to track: 
- [ ] (Optional) If documentation changes are required, which issue on https://github.com/openebs/openebs-docs is used to track them:

rahulkrishnanfs committed:
chart for the dynamic-nfs-provisioner to make the deployment more generic

Signed-off-by: Rahul krishnan R A <rahulkrishnanfs@gmail.com>

rahulkrishnanfs committed:
Added new lines to improve the style

Signed-off-by: Rahul krishnan R A <rahulkrishnanfs@gmail.com>

rahulkrishnanfs committed:
updated missed nfspv reference

Signed-off-by: Rahul krishnan R A <rahulkrishnanfs@gmail.com>

niladrih commented:
@rahulkrishnanfs -- I think we may need the dynamic-localpv-provisioner chart as a dependent chart here because we are using `openebs-hostpath` as the default backend storageClass.

niladrih reviewed (changes requested):
@rahulkrishnanfs -- I think we may need the dynamic-localpv-provisioner chart as a dependent chart here because we are using openebs-hostpath as the default backend storageClass.

niladrih commented:
@rahulkrishnanfs -- Tested this, I got this error from the events of the application pod. Pod remained in ContainerCreating state.
```yaml
Mounting command: systemd-run
Mounting arguments: --description=Kubernetes transient mount for /var/lib/kubelet/pods/6b8d8f85-19cf-4cab-bef5-a7a03e3bbcbf/volumes/kubernetes.io~nfs/pvc-a384f7b7-236a-4508-86a7-6c59359664d8 --scope -- mount -t nfs 10.100.81.23:/ /var/lib/kubelet/pods/6b8d8f85-19cf-4cab-bef5-a7a03e3bbcbf/volumes/kubernetes.io~nfs/pvc-a384f7b7-236a-4508-86a7-6c59359664d8
Output: Running scope as unit: run-r99ee95a41b244866ac61875bf8591202.scope
mount: /var/lib/kubelet/pods/6b8d8f85-19cf-4cab-bef5-a7a03e3bbcbf/volumes/kubernetes.io~nfs/pvc-a384f7b7-236a-4508-86a7-6c59359664d8: bad option; for several filesystems (e.g. nfs, cifs) you might need a /sbin/mount.<type> helper program.
  Warning  FailedMount  30s  kubelet  MountVolume.SetUp failed for volume "pvc-a384f7b7-236a-4508-86a7-6c59359664d8" : mount failed: exit status 32
Mounting command: systemd-run
```

This is the storageClass that got created by default (`kubectl get sc openebs-kernel-nfs -o yaml`)
```yaml
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  annotations:
    cas.openebs.io/config: |
      - name: NFSServerType
        value: kernel
    meta.helm.sh/release-name: nfspv-provisioner
    meta.helm.sh/release-namespace: openebs
    openebs.io/cas-type: nfsrwx
  creationTimestamp: "2021-04-16T16:00:07Z"
  labels:
    app.kubernetes.io/managed-by: Helm
  name: openebs-kernel-nfs
  resourceVersion: "277852"
  selfLink: /apis/storage.k8s.io/v1/storageclasses/openebs-kernel-nfs
  uid: 04befa04-44e8-45b8-aac2-54cffc6bc21b
provisioner: openebs.io/nfsrwx
reclaimPolicy: Delete
volumeBindingMode: Immediate
```
I used the this command to apply the chart when `pwd` was dynamic-nfs-provisioner/deploy/charts/nfspv-provisioner
`helm install -n openebs nfspv-provisioner .`

Might be because of the absence of BackendStorageClass?
cc: @kmova 

EDIT: This is due to https://github.com/openebs/dynamic-nfs-provisioner/issues/16. Testing environment was not configured correctly.

niladrih commented on deploy/charts/nfspv-provisioner/templates/NOTES.txt:
```diff
@@ -0,0 +1,12 @@
+The OpenEBS NFSPV Provisioner has been installed check its status by running:
+$ kubectl get pods -n {{ .Release.Namespace }}
+
+Use `kubectl get bd -n {{ .Release.Namespace }} ` to see the list of
```
This command does not seem relevant to this chart.

quickerdevops replied:
Thanks mate , removed it

rahulkrishnanfs replied:
Removed the mentioned line, thanks

niladrih commented on deploy/charts/nfspv-provisioner/templates/NOTES.txt:
```diff
@@ -0,0 +1,12 @@
+blockdevices attached to the Kubernetes cluster nodes.
+
+For more information, visit our Slack at https://openebs.io/community or view
+the documentation online at http://docs.openebs.io/.
```
NFS provisioner docs are not available at docs.openebs.io at the moment. Could we use https://github.com/openebs/dynamic-nfs-provisioner/blob/develop/README.md instead?

quickerdevops replied:
Pointed to the developer readme file as you suggested , Thanks

rahulkrishnanfs replied:
Thanks mate , Updated to the developer readme file as you suggested

niladrih reviewed (changes requested):
@rahulkrishnanfs -- My testing environment was configured incorrectly while testing, as mentioned in the previous comment. 

The chart looks good. The directory structure followed could be changed to the `deploy/helm/charts` instead of `deploy/charts`.

Suggested some chart deploy message changes.

rahulkrishnanfs committed:
Merge branch 'develop' into 15-dynamic-nfs-provisioner-helm-chart

rahulkrishnanfs committed:
Fix in the chart notes and moved the charts under helm directory

Signed-off-by: Rahul krishnan R A <rahulkrishnanfs@gmail.com>

quickerdevops commented:
> @rahulkrishnanfs -- My testing environment was configured incorrectly while testing, as mentioned in the previous comment.
> 
> The chart looks good. The directory structure followed could be changed to the `deploy/helm/charts` instead of `deploy/charts`.
> 
> Suggested some chart deploy message changes.

@niladrih  Thanks mate for the suggestion. I have changed the directory structure as you suggested

rahulkrishnanfs commented:
> @rahulkrishnanfs -- My testing environment was configured incorrectly while testing, as mentioned in the previous comment.
> 
> The chart looks good. The directory structure followed could be changed to the `deploy/helm/charts` instead of `deploy/charts`.
> 
> Suggested some chart deploy message changes.

@niladrih Thanks mate for the suggestion. I have changed the directory structure as you suggested

niladrih commented on deploy/helm/charts/nfspv-provisioner/README.md:
```diff
@@ -0,0 +1,93 @@
+| Parameter                                   | Description                                   | Default                                   |
+| ------------------------------------------- | --------------------------------------------- | ----------------------------------------- | 
+| `analytics.enabled`                         | Enable sending stats to Google Analytics          | `true`                          |
+| `imagePullSecrets`                          | Provides image pull secrect                       | `""`                            |
```
```suggestion
| `imagePullSecrets`                          | Provides image pull secret                       | `""`                            |
```

niladrih commented on deploy/helm/charts/nfspv-provisioner/README.md:
```diff
@@ -0,0 +1,93 @@
+| `nfspv.enabled`                             | Enable NFSPV Provisioner                          | `true`                          |
+| `nfspv.image.registry`                      | Registry for NFSPV Provisioner image              | `""`                            |
+| `nfspv.image.repository`                    | Image repository for NFSPV Provisioner            | `openebs/provisioner-nfs-amd64` |
+| `nfspvpv.image.tag`                         |	Image tag for NFSPV Provisioner	                  | `0.2.0`                         |
```
```suggestion
| `nfspv.image.tag`                           | Image tag for NFSPV Provisioner	                  | `0.2.0`                         |
```

mynktl replied:
typo: `nfspvpv` -> `nfspv`

niladrih commented on deploy/helm/charts/nfspv-provisioner/templates/kernel-nfs-storageclass.yaml:
```diff
@@ -0,0 +1,12 @@
+    openebs.io/cas-type: nfsrwx
+    cas.openebs.io/config: |
+      - name: NFSServerType
+        value: {{ .Values.nfsStorageClass.nfsServerType }}
```
```suggestion
        value: {{ .Values.nfsStorageClass.nfsServerType }}
{{- if .Values.nfsStorageClass.BackendStorageClass }}
      - name: BackendStorageClass
        value: {{ .Values.nfsStorageClass.BackendStorageClass }}
{{- end }}
```
Can we do something like this, so that the user can specify the backend storageClass with a flag like `--set nfsStorageClass.BackendStorageClass=<storageClass>` ?

kmova replied:
```suggestion
        value: {{ .Values.nfsStorageClass.nfsServerType }}
{{- if .Values.nfsStorageClass.backendStorageClass }}
      - name: BackendStorageClass
        value: {{ .Values.nfsStorageClass.backendStorageClass }}
{{- end }}
```

niladrih commented on deploy/helm/charts/nfspv-provisioner/values.yaml:
```diff
@@ -0,0 +1,72 @@
+nfsStorageClass:
+  name: openebs-kernel-nfs
+  reclaimPolicy: Delete
+  nfsServerType: kernel
```
```suggestion
  nfsServerType: kernel
  BackendStorageClass: ""
```
Could we add the option here to add a BackendStorageClass?

kmova replied:
```suggestion
  nfsServerType: kernel
  backendStorageClass: ""
```

mynktl commented on deploy/helm/charts/nfspv-provisioner/values.yaml:
```diff
@@ -0,0 +1,72 @@
+    # Make sure that registry name end with a '/'.
+    # For example : quay.io/ is a correct value here and quay.io is incorrect
+    registry:
+    repository: openebs/provisioner-nfs-amd64
```
registry should be `openebs/` and repository `provisioner-nfs-amd64`

kmova replied:
```suggestion
    repository: openebs/provisioner-nfs
```

kmova commented on deploy/helm/charts/nfspv-provisioner/Chart.yaml:
```diff
@@ -0,0 +1,29 @@
+apiVersion: v2
+name: nfspv-provisioner
+description: Helm chart for OpenEBS Dynamic NFS PV. For instructions to install OpenEBS Dynamic NFS PV using helm chart, refer to https://github.com/openebs/dynamic-nfs-provisioner.
```
```suggestion
description: Helm chart for OpenEBS Dynamic NFS PV. For instructions to install OpenEBS Dynamic NFS PV using helm chart, refer to https://openebs.github.io/dynamic-nfs-provisioner.
```

kmova commented on deploy/helm/charts/nfspv-provisioner/templates/NOTES.txt:
```diff
@@ -0,0 +1,9 @@
+$ kubectl get pods -n {{ .Release.Namespace }}
+
+For more information, visit our Slack at https://openebs.io/community or view
+the documentation online at https://github.com/openebs/dynamic-nfs-provisioner/blob/develop/develop/developer.md .
```
```suggestion
the documentation online at https://github.com/openebs/dynamic-nfs-provisioner/.
```

kmova commented on deploy/helm/charts/nfspv-provisioner/values.yaml:
```diff
@@ -0,0 +1,72 @@
+  image:
+    # Make sure that registry name end with a '/'.
+    # For example : quay.io/ is a correct value here and quay.io is incorrect
+    registry:
```
```suggestion
    registry: openebs
```

Commit:
Apply suggestions from code review

Co-authored-by: Niladri Halder <55788661+niladrih@users.noreply.github.com>

kmova commented on deploy/helm/charts/nfspv-provisioner/Chart.yaml:
```diff
@@ -0,0 +1,28 @@
+apiVersion: v2
+name: nfspv-provisioner
```
```suggestion
name: nfs-provisioner
```

kmova commented on deploy/helm/charts/nfspv-provisioner/README.md:
````diff
@@ -0,0 +1,93 @@
+## Get Repo Info
+
+```console
+helm repo add openebs-nfspv https://openebs.github.io/dynamic-nfspv-provisioner
````
```suggestion
helm repo add openebs-nfs https://openebs.github.io/dynamic-nfs-provisioner
```

kmova commented on deploy/helm/charts/nfspv-provisioner/README.md:
````diff
@@ -0,0 +1,93 @@
+
+```console
+# Helm
+$ helm install [RELEASE_NAME] openebs-nfspv/nfspv-provisioner
````
```suggestion
$ helm install [RELEASE_NAME] openebs-nfs/nfs-provisioner
```

kmova commented on deploy/helm/charts/nfspv-provisioner/README.md:
````diff
@@ -0,0 +1,93 @@
+Alternatively, a YAML file that specifies the values for the parameters can be provided while installing the chart. For example,
+
+```bash
+helm install <release-name> -f values.yaml ----namespace openebs nfspv-provisioner
````
```suggestion
helm install <release-name> -f values.yaml ----namespace openebs nfs-provisioner
```

niladrih reviewed (approved):
Looks good to me. Some changes will be taken up in a separate PR.

kmova commented on deploy/helm/charts/nfspv-provisioner/values.yaml:
```diff
@@ -0,0 +1,73 @@
+  image:
+    # Make sure that registry name end with a '/'.
+    # For example : quay.io/ is a correct value here and quay.io is incorrect
+    registry: openebs
```
```suggestion
    registry: 
```

Commit:
Update deploy/helm/charts/nfspv-provisioner/values.yaml

incorrectly set to openebs. This should be either quay or other repos. By default it will be docker.

niladrih reviewed (approved):
LGTM
